test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/BodyModelBinderTests.cs

```diff
@@ -206,6 +206,60 @@ public class BodyModelBinderTests
+            {
+                new TestInputFormatter(canRead: false),
+                new TestInputFormatter(canRead: true),
+            };
+
+            var provider = new TestModelMetadataProvider();
+            provider.ForType<Person>().BindingDetails(d => d.BindingSource = BindingSource.Body);
+            var bindingContext = GetBindingContext(typeof(Person), metadataProvider: provider);
+            bindingContext.HttpContext.Request.ContentType = "application/json";
+            var binder = new BodyModelBinder(inputFormatters, new TestHttpRequestStreamReaderFactory(), loggerFactory);
+
+            // Act
+            await binder.BindModelAsync(bindingContext);
+
+            // Assert
+            Assert.Equal($"Rejected input formatter 'Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinderTests+TestInputFormatter' for content type 'application/json'", sink.Writes[0].State?.ToString());
+            Assert.Equal($"Selected input formatter 'Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinderTests+TestInputFormatter' for content type 'application/json'", sink.Writes[1].State?.ToString());
+        }
+
+        [Fact]
+        public async Task BindModelAsync_LogsNoFormatterSelectedAndRemoveFromBodyAttribute()
+        {
+            // Arrange
+            var sink = new TestSink();
+            var loggerFactory = new TestLoggerFactory(sink, enabled: true);
+            var inputFormatters = new List<IInputFormatter>()
+            {
+                new TestInputFormatter(canRead: false),
+                new TestInputFormatter(canRead: false),
+            };
+
+            var provider = new TestModelMetadataProvider();
+            provider.ForType<Person>().BindingDetails(d => d.BindingSource = BindingSource.Body);
+            var bindingContext = GetBindingContext(typeof(Person), metadataProvider: provider);
+            bindingContext.HttpContext.Request.ContentType = "multipart/form-data";
+            var binder = new BodyModelBinder(inputFormatters, new TestHttpRequestStreamReaderFactory(), loggerFactory);
+
+            // Act
+            await binder.BindModelAsync(bindingContext);
+
+            // Assert
+            Assert.Equal($"Rejected input formatter 'Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinderTests+TestInputFormatter' for content type 'multipart/form-data'", sink.Writes[0].State?.ToString());
+            Assert.Equal($"Rejected input formatter 'Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinderTests+TestInputFormatter' for content type 'multipart/form-data'", sink.Writes[1].State?.ToString());
+            Assert.Equal($"No input formatter was found to support the content type 'multipart/form-data' for use with the [FromBody] attribute.", sink.Writes[2].State?.ToString());
+            Assert.Equal($"To use model binding, remove the [FromBody] attribute from the action method's parameter.", sink.Writes[3].State?.ToString());
```

Comment:
Suggest `Assert.Collection()` might make these series of assertions easier to read e.g.
``` c#
Assert.Collection(
    sink.Writes,
    write => Assert.Equal(
        "Rejected input formatter 'Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinderTests+TestInputFormatter' for content type 'multipart/form-data'",
        write.State.ToString()),
    // ...
    );
```